app/jobs/autoscaling.py

```diff
@@ -198,9 +208,146 @@ def set_desired_capacity(asg, compute_desired_instances, can_scale=True):
     if can_scale:
         if DEBUG:
             print cmd
+        if num_desired < previous_desired:
+            start_draining(asg, previous_desired - num_desired)
+        elif num_desired > previous_desired:
+            stop_draining(asg, num_desired - previous_desired)
         aws_command(cmd)
 
 
+def instances_in(asg):
+    return [item["InstanceId"] for item in asg["Instances"]]
+
+
+def unixtime_now():
+    return int(time.time())
+
+
+def iso2unixtime(iso_str):
+    return int(dateutil.parser.parse(iso_str).strftime('%s'))
+
+
+def add_draining_tag(instance_ids):
+    cmd = "aws ec2 create-tags --resources {list_instances} --tags Key={draining_tag},Value={timestamp}"
+    cmd = cmd.format(list_instances=' '.join(instance_ids), draining_tag=DRAINING_TAG, timestamp=unixtime_now())
+    aws_command(cmd)
+
+
+def remove_draining_tag(instance_ids):
+    delete_tags_from_instances(instance_ids, [DRAINING_TAG])
+
+
+def start_draining(asg, num_instances):
+    instance_ids = instances_to_drain(asg, num_instances)
+    add_draining_tag(instance_ids)
+
+
+def stop_draining(asg, num_instances):
+    instance_ids = instances_to_rescue(asg, num_instances)
+    if instance_ids:
+        print "Stopping drainage of the following instances: " + ",".join(instance_ids)
+        add_termination_protection(instance_ids, asg)
+        # Note: if autoscaling.py ran too frequently, there would be a risk that an instance enters Terminating state between the time
+        # of the health check in instances_to_rescue and the time of add_termination_protection, in which case add_termination_protection
+        # would throw an error. But because autoscaling.py runs only infrequently, we can be sure that there is no scaling decision in limbo
+        # from the last execution of autoscaling.py.
```

Comment:
Hmm, no, I think that won't happen, because we are taking care to never rescue from draining an instance whose termination protection has been removed.

Reply:
Also in that design as we discussed in person we don't have to re-add termination protection.